src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+#include "umpire/event/event.hpp"
+#include "umpire/strategy/PoolCoalesceHeuristic.hpp"
+#include "umpire/strategy/mixins/AlignedAllocation.hpp"
+#include "umpire/util/Macros.hpp"
+#include "umpire/util/memory_sanitizers.hpp"
+
+namespace umpire {
+namespace strategy {
+
+ResourceAwarePool::ResourceAwarePool(const std::string& name, int id, Allocator allocator,
+                                     const std::size_t first_minimum_pool_allocation_size,
+                                     const std::size_t next_minimum_pool_allocation_size, std::size_t alignment,
+                                     PoolCoalesceHeuristic<ResourceAwarePool> should_coalesce) noexcept
+    : AllocationStrategy{name, id, allocator.getAllocationStrategy(), "ResourceAwarePool"},
+      mixins::AlignedAllocation{alignment, allocator.getAllocationStrategy()},
+      m_should_coalesce{should_coalesce},
+      m_first_minimum_pool_allocation_size{first_minimum_pool_allocation_size},
+      m_next_minimum_pool_allocation_size{next_minimum_pool_allocation_size}
+{
+  UMPIRE_LOG(Debug, " ( "
+                        << "name=\"" << name << "\""
+                        << ", id=" << id << ", allocator=\"" << allocator.getName() << "\""
+                        << ", first_minimum_pool_allocation_size=" << m_first_minimum_pool_allocation_size
+                        << ", next_minimum_pool_allocation_size=" << m_next_minimum_pool_allocation_size
+                        << ", alignment=" << alignment << " )");
+}
+
+ResourceAwarePool::~ResourceAwarePool()
+{
+  UMPIRE_LOG(Debug, "Releasing free blocks to device");
+  m_is_destructing = true;
+  release();
+}
+
+void* ResourceAwarePool::allocate(std::size_t UMPIRE_UNUSED_ARG(bytes))
+{
+  void* ptr{nullptr};
+  UMPIRE_ERROR(
+      runtime_error,
+      fmt::format("The ResourceAwarePool requires a Camp resource. See "
+                  "https://umpire.readthedocs.io/en/develop/sphinx/cookbook/resource_aware_pool.html for more info."));
+  return ptr;
+}
+
+void* ResourceAwarePool::allocate_resource(camp::resources::Resource r, std::size_t bytes)
+{
+  UMPIRE_LOG(Debug, "(bytes=" << bytes << ")");
+  const std::size_t rounded_bytes{aligned_round_up(bytes)};
+  const auto& best = m_free_map.lower_bound(rounded_bytes);
+
+  Chunk* chunk{nullptr};
+
+  if (!m_pending_map.empty()) {
+    for (auto it = m_pending_map.begin(); it != m_pending_map.end();) {
+      auto pending_chunk = (*it);
+
+      if (pending_chunk->size >= rounded_bytes && pending_chunk->m_resource == r) { // reusing chunk with same resource
+        chunk = pending_chunk;
+        chunk->m_resource = pending_chunk->m_resource;
+        chunk->m_event = pending_chunk->m_event;
+        chunk->free = false;
+        m_pending_map.erase(it);
+        break;
+      }
+
+      if (pending_chunk->free == false && pending_chunk->m_event.check()) // reuse no-longer-pending chunk
+      {
+        do_deallocate(pending_chunk, pending_chunk->data);
```

Comment:
Could you pass the iterator to do_deallocate so you don't have to search pending again for it?